tools/sgx-capability/check-sgx-capability.c

```diff
@@ -0,0 +1,106 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Copyright (C) 2011-2019 Intel Corporation. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in
+ *     the documentation and/or other materials provided with the
+ *     distribution.
+ *   * Neither the name of Intel Corporation nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include <stdio.h>
+#include <sgx_capable.h>
+
+int main (int argc, char *argv[])
+{
+	sgx_status_t status;
+	sgx_device_status_t result;
+
+	status= sgx_cap_get_status(&result);
+	if (status != SGX_SUCCESS) {
+		switch(status) {
+		case SGX_ERROR_NO_PRIVILEGE:
+			fprintf(stderr, "could not examine the EFI filesystem\n");
+			return 1;
+		case SGX_ERROR_INVALID_PARAMETER:
+		case SGX_ERROR_UNEXPECTED:
+			fprintf(stderr, "could not get SGX status: ");
+		default:
+			fprintf(stderr, "sgx_cap_get_status returned 0x%04x\n", status);
+			return 1;
+		}
+	}
+
+	if (result == SGX_ENABLED) {
+		printf("Intel SGX is supported and enabled on this system.\n");
+		return 0;
+	} else if (result == SGX_DISABLED_UNSUPPORTED_CPU) {
+		printf("This CPU does not support Intel SGX.\n");
+		return 1;
+	} else if (result == SGX_DISABLED_LEGACY_OS) {
+		printf("This processor supports Intel SGX but was booted in legacy mode.\n");
+		printf("A UEFI boot is required to determine whether or not your BIOS\n");
+		printf("supports Intel SGX, and to attempt the software enable.\n");
+		printf("Alternatively, you can check your BIOS to see if it has an\n");
+		printf("explicit option to enable SGX.\n");
```

Comment:
Maybe we should not mention software enable in the output as it might be a bit confusing? I suggest the following output text:

```
This processor supports Intel SGX but was booted in legacy mode.
A UEFI boot is required to determine whether or not your BIOS
supports Intel SGX.
```

Reply:
Thanks, I'll update this message.